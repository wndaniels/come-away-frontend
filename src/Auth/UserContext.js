import React from "react";
/**
 *  Context:
 *
 * Provides currentUser object and setter for it throughout app.
 */
const UserContext = React.createContext();

export default UserContext;
