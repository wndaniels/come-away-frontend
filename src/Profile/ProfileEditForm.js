import { useState, useContext, useEffect } from "react";
import UserContext from "../Auth/UserContext";
import { Form, useParams, useNavigate } from "react-router-dom";
import Alert from "../Common/Alert";
import ComeAwayApi from "../api/api";

const ProfileForm = () => {
  const navigate = useNavigate();
  const { currentUser, setCurrentUser } = useContext(UserContext);
  const { username } = useParams(currentUser.username);

  useEffect(function getCurrentUsername() {});
  const [formData, setFormData] = useState({
    firstName: currentUser.firstName,
    lastName: currentUser.lastName,
    email: currentUser.email,
    username: currentUser.username,
  });

  const [formError, setFormError] = useState([]);
  const [savedEdit, setSavedEdit] = useState(false);

  async function handleSubmit(evt) {
    evt.preventDefault();

    let profileData = {
      firstName: formData.firstName,
      lastName: formData.lastName,
      email: formData.email,
    };

    let username = formData.username;
    let updateUser;

    try {
      updateUser = await ComeAwayApi.saveProfile(username, profileData);
    } catch (errors) {
      setFormError(errors);
      return;
    }

    navigate("/");
    setFormData((f) => ({ ...f }));
    setFormError([]);
    setSavedEdit(true);

    setCurrentUser(updateUser);
  }

  function handleChange(evt) {
    const { name, value } = evt.target;
    setFormData((f) => ({
      ...f,
      [name]: value,
    }));
    setFormError([]);
  }

  return (
    <div className="ProfileEditForm">
      <div className="container col-md-6 offset-md-3 col-lg-4 offset-lg-4">
        <h1 className="mb-3">Profile Form</h1>
        <div className="card">
          <div className="card-body">
            <Form onSubmit={handleSubmit}>
              <div className="d-grid gap-2">
                <div className="form-group">
                  <h3>{formData.username}</h3>
                </div>
                <div className="form-group">
                  <label className="mb-1">First Name:</label>
                  <input
                    name="firstName"
                    className="form-control"
                    value={formData.firstName}
                    onChange={handleChange}
                  />
                </div>
                <div className="form-group">
                  <label className="mb-1">Last Name:</label>
                  <input
                    name="lastName"
                    className="form-control"
                    value={formData.lastName}
                    onChange={handleChange}
                  />
                </div>
                <div className="form-group">
                  <label className="mb-1">Email:</label>
                  <input
                    name="email"
                    className="form-control mb-3"
                    value={formData.email}
                    onChange={handleChange}
                  />
                </div>
              </div>

              {formError.length ? (
                <Alert type="danger" messages={formError} />
              ) : null}

              {savedEdit ? (
                <Alert type="success" messages={["Updated successfully."]} />
              ) : null}

              <button className="btn btn-sm btn-primary float-right">
                Save Changes
              </button>
            </Form>
          </div>
        </div>
      </div>
    </div>
  );
};

export default ProfileForm;
