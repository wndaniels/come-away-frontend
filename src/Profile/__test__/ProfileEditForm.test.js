import React from "react";
import { getByRole, render } from "@testing-library/react";
import "@testing-library/jest-dom/extend-expect";
import "@testing-library/jest-dom";
import ProfileEditForm from "../ProfileEditForm";
import UserContext from "../../Auth/UserContext";

jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"),
  useNavigate: () => jest.fn(),
}));

describe("ProfileEditForm", () => {
  it("renders without crashing", () => {
    const { asFragment } = render(
      <Wrapper>
        <ProfileEditForm />
      </Wrapper>
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it("can find heading", () => {
    const { getByRole } = render(
      <Wrapper>
        <ProfileEditForm />
      </Wrapper>
    );

    const heading = getByRole("heading", {
      name: /profile form/i,
      class: "mb-3",
    });

    expect(heading).toBeInTheDocument();
  });
});

function Wrapper({ children }) {
  return (
    <UserContext.Provider
      value={{
        currentUser: jest.fn(),
        setCurrentUser: jest.fn(),
        login: jest.fn(),
        signup: jest.fn(),
        logout: jest.fn(),
      }}
    >
      {children}
    </UserContext.Provider>
  );
}
