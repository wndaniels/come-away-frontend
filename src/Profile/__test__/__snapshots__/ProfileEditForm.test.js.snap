// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProfileEditForm renders without crashing 1`] = `
<DocumentFragment>
  <div
    class="ProfileEditForm m-5"
  >
    <div
      class="container col-md-6 offset-md-3 col-lg-4 offset-lg-4"
    >
      <h1
        class="mb-3"
      >
        Profile Form
      </h1>
      <div
        class="card"
      >
        <div
          class="card-body"
        >
          <form>
            <div
              class="d-grid gap-2"
            >
              <div
                class="form-group"
              >
                <h3 />
              </div>
              <div
                class="form-group"
              >
                <label
                  class="mb-1"
                >
                  First Name:
                </label>
                <input
                  class="form-control"
                  name="firstName"
                  value=""
                />
              </div>
              <div
                class="form-group"
              >
                <label
                  class="mb-1"
                >
                  Last Name:
                </label>
                <input
                  class="form-control"
                  name="lastName"
                  value=""
                />
              </div>
              <div
                class="form-group"
              >
                <label
                  class="mb-1"
                >
                  Email:
                </label>
                <input
                  class="form-control mb-3"
                  name="email"
                  value=""
                />
              </div>
            </div>
            <button
              class="btn btn-sm btn-primary float-right"
            >
              Save Changes
            </button>
          </form>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
